lib/util/itinerary.js

```diff
@@ -245,31 +245,133 @@ export function getLegBounds (leg) {
+ * object has undefined, null or an empty string, but the other does have a
+ * string with length > 0, then that string will get priority.
+ *
+ * @param  {string} key The key from which to obtain the sort comparison value
+ */
+function makeObjectStringValueComparator (key) {
+  return (a, b) => {
+    const aVal = a[key]
+    const bVal = b[key]
+    // both a and b are uncomparable strings, return equivalent value
+    if (!aVal && !bVal) return 0
+    // a is not a comparable string, b gets priority
+    if (!aVal) return 1
+    // b is not a comparable string, a gets priority
+    if (!bVal) return -1
+    // a and b are comparable strings, return the sort value
+    if (aVal < bVal) return -1
+    if (aVal > bVal) return 1
+    return 0
+  }
+}
+
+/**
+ * OpenTripPlanner sets the routeSortOrder to -999 by default. So, if that value
+ * is encountered, assume that it actually means that the routeSortOrder is not
+ * set in the GTFS.
+ *
+ * See https://github.com/opentripplanner/OpenTripPlanner/issues/2938
+ * Also see https://github.com/opentripplanner/otp-react-redux/issues/122
+ */
+function getRouteSortOrderValue (val) {
+  return val === -999 ? undefined : val
+}
+
+/**
+ * Create a multi-criteria sort comparator function composed of other sort
+ * comparator functions. Each comparator function will be ran in the order given
+ * until a non-zero comparison value is obtained which is then immediately
+ * returned. If all comparison functions return equivalance, then the values
+ * are assumed to be equivalent.
+ */
+function makeMultiCriteriaSort (...criteria) {
+  return (a, b) => {
+    for (let i = 0; i < criteria.length; i++) {
+      const curCriteriaComparatorValue = criteria[i](a, b)
+      // if the comparison objects are not equivalent, return the value obtained
+      // in this current criteria comparison
+      if (curCriteriaComparatorValue !== 0) {
+        return curCriteriaComparatorValue
+      }
+    }
+    return 0
   }
-  if (aComp < bComp) return -1
-  if (aComp > bComp) return 1
-  return 0
 }
 
+/**
+ * Compares routes for the purposes of sorting and displaying in a user
+ * interface. Due to GTFS feeds having varying levels of data quality, a
+ * multi-criteria sort is needed to account for various differences. The sort
+ * occurs as follows:
+ * 1. Attempt to sort based off of sortOrder. Routes that do not have a valid
+ *    sortOrder will be placed beneath those that do.
+ * 2. In case of the same sortOrder criteria, attempt to sort based off of
+ *    shortNames that are parseable as integers. shortNames that cannot be
+ *    parsed as integers will be placed beneath those that are valid.
+ * 3. In case of the same integer shortName criteria, attempt to sort based
+ *    off of the string shortName. Routes without shortNames will be placed
+ *    beneath those with shortNames.
+ * 4. In case of the same string shortName, attempt to sort based off of
+ *    the string longName.
```

Comment:
The way this list of sorting criteria is worded is somewhat confusing to me. Instead of nestling the different cases/conditions in each item in the comment, I think it would be more helpful to state that when a criteria is found to be equivalent for a pair of routes, the next criteria is used. Then just list the criteria being used. This is somewhat of a restatement of the `makeMultiCriteriaSort` jsdoc, but I think it bears repeating. How I might rewrite it:

```
Compares routes for the purposes of sorting and displaying in a user interface. Due to GTFS feeds having varying levels of data quality, a multi-criteria sort is needed to account for various differences. The criteria included here are each applied to the routes in the order listed. If a given sort criterion yields equivalence (e.g., two routes have the short name "20"), the comparator falls back onto the next sort criterion (e.g., long name). The sort occurs as follows:
 1. sortOrder. Routes that do not have a valid sortOrder will be placed beneath those that do.
 2. shortName as integer. shortNames that cannot be parsed as integers will be placed beneath those that are valid.
 3. shortName as string. Routes without shortNames will be placed beneath those with shortNames.
 4. longName as string.
```